### DISCUSSION TITLE
In an Object Oriented design, how do you define an endpoints RequestModel at runtime?

### DISCUSSION THREAD

**Discussion by manuel-wagesreither** [fastapi/fastapi#11103] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter
from pydantic import BaseModel, Field
from typing import Literal, Union


class MediaPipeline1():
    class Params1(BaseModel):
        pipeline_name: Literal["One"]
        foo: int = 11

class MediaPipeline2():
    class Params2(BaseModel):
        pipeline_name: Literal["Two"]
        bar: int = 22

class MediaPipeline3():
    class Params3(BaseModel):
        pipeline_name: Literal["Three"]
        baz: int = 33

##############################################

class MediaController():
    # Hardcode the dependencies the initializer gets passed
    # This is bad
    class RequestModel(BaseModel): # Tagged union
        params: Union[MediaPipeline1.Params1, MediaPipeline2.Params2] = Field(discriminator='pipeline_name')

    def __init__(self, pipelines):
        self.pipelines = pipelines

    def handler(self, request: RequestModel):
        print(request)

#############################################

my_pipeline_1 = MediaPipeline1()
my_pipeline_2 = MediaPipeline2()
my_pipeline_3 = MediaPipeline3()

router = APIRouter()

# Works fine
my_controller_12 = MediaController([my_pipeline_1, my_pipeline_2])
router.add_api_route("/1", my_controller_12.handler)

# Bug: Will not handle my_pipeline_3 because class is hardcoded for [my_pipeline_1, my_pipeline_2]
my_controller_23 = MediaController([my_pipeline_2, my_pipeline_3])
router.add_api_route("/2", my_controller_23.handler)

app = FastAPI()
app.include_router(router)
```

### Description

I'd like to loose the coupling between `MediaController` and its dependencies, the `MediaPipelineX`s to enable it to work with whatever I'm passing to it. How do I do that? I feel if [`add_api_route`](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L843) would take another parameter `request_model` I could create the type in the class initializer and provide it as instance attribute.

### Operating System

Linux

### Operating System Details

Debian GNU/Linux 12 (bookworm)

### FastAPI Version

0.99.1

### Pydantic Version

1.10.11

### Python Version

Python 3.11.2

**Comment by pbasista** [comment]
> I'd like to loose the coupling between MediaController and its dependencies, the MediaPipelineXs to enable it to work with whatever I'm passing to it.

Could you describe, in generic language, without referring to the models in your example, the desired functionality? Perhaps the presented approach is unsuitable for your desired goal and maybe there is a much simpler way to do it.

**Reply by manuel-wagesreither** [reply]
I'd like to create a bunch of objects that shall be available over an API. Every object shall have it's own API endpoint. For validating the data that is passed to the endpoint, I'd like to leverage Pydantic. Each endpoint shall have a different validation model. It can be constructed based on what gets passed into the objects initializer.

If I initialize the object with three objects (= dependencies) `dep1`, `dep2` `dep3`, the endpoint validation model shall be  `Union[dep1.ValidationModel, dep2.ValidationModel, dep3.ValidationModel]`.

**Reply by pbasista** [reply]
> Every object shall have it's own API endpoint.
> Each endpoint shall have a different validation model.

Ok. So, if I understand correctly, you are trying to achieve something like this:

1. endpoint `/dogs` has a model, e.g. `Dog`
2. endpoint `/cats` has a different model, e.g. `Cat`
3. ... and so on

> It can be constructed based on what gets passed into the objects initializer.

I am unsure what you mean. Are you suggesting to dynamically select the model based on some endpoint-specific or user-provided data? Like, for instance, based on the path component of the URL? For example, when using the `/dogs` path, use the `Dog` model? If so, where should such a mapping be defined and how?

> ... the endpoint validation model shall be `Union[...`

In general, defining type as a union means that any of the provided types can be used. A particular type would not be tied to the selection criteria (like the URL path) that you mentioned. Instead, all the types within the union will always be allowed by FastAPI.

Are you aiming to define only one route handler function and expect it to handle all of these URL paths with different validation criteria for different values within the URL, for example? That would mean to dynamically select a model to validate at runtime. You can do that manually within the handler of course. Check the path. Select the model. Validate. But I am unsure if FastAPI can be used to orchestrate something like that for you.

**Reply by manuel-wagesreither** [reply]
Sorry for taking so long the respond – I've been abroad.

I'd like to create a class `AnimalEndpoint`, that, when instantiated with `list(Dog,Cat)`, accepts the following input:

```
{
    species: string = "Cat"
    meows: bool
}
```
or
```
{
    species: string = "Dog"
    barks: bool
}
```
The OpenAPI spec shall be created accordingly.
If the input is neither `Cat` nor `Dog`, FastAPI shall throw a validation error.

However, `AnimalEndpoint` is a class. Hence it can be instantiated numerous times. So, when creating a second instance, this time with passing a `list(Dog,Duck)` to the initializer, it shall accept the following:

```
{
    species: string = "Dog"
    barks: bool
}
```
or
```
{
    species: string = "Duck"
    quacks: bool
}
```
Again, OpenAPI spec shall be created accordingly and an error be thrown when the input is neither a Dog or a Duck.

In the example I posted at the very beginning, `MediaPipeline1`, -`2` and -`3` can be replaced by `Cat`, `Dog` and `Duck`. `AnimalEndpoint` would be `MediaController`.

In the meantime I'm fairly sure the root cause of my problem is that FastAPI takes the validation model exclusively from the handler functions type hints, which, if the handler function is a method, ultimately is a class attribute.

Hence, all objects created from that class *must* share the same validation model.

That is unfortunate, because it doesn't seem to be the case for the response model. This can be set not on a per-class, but on a [per-endpoint level](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L848).

**Reply by manuel-wagesreither** [reply]
> That would mean to dynamically select a model to validate at runtime. You can do that manually within the handler of course. Check the path. Select the model. Validate. But I am unsure if FastAPI can be used to orchestrate something like that for you.

I'd really like to have FastAPI create the correct OpenAPI spec. :-(

But if I ultimately fail doing that: How would I validate the model at runtime and emit an HTTP error if validation fails? Ideally, it pass the validation error as reply to the HTTP client just the same way as if FastAPI would do the validation.

Btw, thanks for your support @pbasista! Much appreciated.

**Reply by pbasista** [reply, marked as answer]
> I'd really like to have FastAPI create the correct OpenAPI spec

I see. In such case perhaps consider creating a function which takes the type of the handler as a parameter and returns a function that will have the correct type hint and could be used as a FastAPI handler. It might look similar to this:

```python
T = TypeVar("T")

def get_generic_handler(T, path: str):

    @api.post(path)
    def handler(body: T) -> str:
        return f"{path}: Hello {body.name}"

    return handler

cats_and_dogs = get_generic_handler(Cat | Dog, "/cats_and_dogs")
```

**Reply by manuel-wagesreither** [reply]
This works. I was so focused on altering the type hints of an existing method, that the idea of creating the entire function at runtime didn't cross my mind.

Thank you again! :bowing_man:
